### PR TITLE
Refactor the hyperkube image and add various features

### PR DESCRIPTION
**Pull request by luxas**
@cheld @zreigz @mikedanese

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30299)

### CODE REVIEW

**Comment by zreigz** [comment]
A lot of nice improvements. I am curious about  static-pods. What is the best approach to test it with hypercube. Yesterday I've tested it manually. I took custom hypercube with CNI. I mounted new directory where I copied json files  from here: https://github.com/kubernetes/kubernetes/tree/master/cluster/images/hyperkube/static-pods and I added the path to this parameter `--config=<directory>` Is any other way to enable static-pods in hypercube ?

File: cluster/images/hyperkube/Makefile [hunk 1]
```diff
@@ -54,33 +54,35 @@ ifndef VERSION
...
 	cd ${TEMP_DIR} && sed -i.back "s|ARCH|${QEMUARCH}|g" Dockerfile
 	cd ${TEMP_DIR} && sed -i.back "s|BASEIMAGE|${BASEIMAGE}|g" Dockerfile
-	cd ${TEMP_DIR} && sed -i.back "s|-amd64|-${ARCH}|g" addons/*.yaml
-	cd ${TEMP_DIR} && sed -i.back "s|__PILLAR__DNS__REPLICAS__|1|g;s|__PILLAR__DNS__SERVER__|10.0.0.10|g;" addons/skydns*.yaml
-	cd ${TEMP_DIR} && sed -i.back "s|__PILLAR__DNS__DOMAIN__|cluster.local|g;s|__PILLAR__FEDERATIONS__DOMAIN__MAP__||g;" addons/skydns*.yaml
-	rm ${TEMP_DIR}/addons/*.back
+	cd ${TEMP_DIR} && sed -i.back "s|-amd64|-${ARCH}|g" addons/singlenode/*.yaml addons/multinode/*.yaml
+	cd ${TEMP_DIR} && sed -i.back "s|__PILLAR__DNS__REPLICAS__|1|g;s|__PILLAR__DNS__SERVER__|10.0.0.10|g;" addons/singlenode/skydns*.yaml addons/multinode/skydns*.yaml
+	cd ${TEMP_DIR} && sed -i.back "s|__PILLAR__DNS__DOMAIN__|cluster.local|g;s|__PILLAR__FEDERATIONS__DOMAIN__MAP__||g;" addons/singlenode/skydns*.yaml addons/multinode/skydns*.yaml
+	cd ${TEMP_DIR} && rm -f addons/singlenode/*.back addons/multinode/*.back
```
Comment (cheld) [inline cluster/images/hyperkube/Makefile]: missing cleanup: static-pods/*.back

File: cluster/images/hyperkube/static-pods/master-multi.json [hunk 2]
```diff
@@ -57,18 +59,19 @@
...
       "name": "setup",
       "image": "REGISTRY/hyperkube-ARCH:VERSION",
       "command": [
-              "/setup-files.sh",
-              "IP:10.0.0.1,DNS:kubernetes,DNS:kubernetes.default,DNS:kubernetes.default.svc,DNS:kubernetes.default.svc.cluster.local"
+        "/setup-files.sh",
+        "IP:10.0.0.1,DNS:kubernetes,DNS:kubernetes.default,DNS:kubernetes.default.svc,DNS:kubernetes.default.svc.cluster.local"
       ],
       "volumeMounts": [
         {
```
Comment (cheld) [inline cluster/images/hyperkube/static-pods/master-multi.json]: use "mountPath": "/srv/kubernetes"

File: cluster/images/hyperkube/static-pods/master-multi.json [hunk 3]
```diff
@@ -12,13 +12,15 @@
...
-              "/hyperkube",
-              "controller-manager",
-              "--master=127.0.0.1:8080",
-              "--service-account-private-key-file=/srv/kubernetes/server.key",
-              "--root-ca-file=/srv/kubernetes/ca.crt",
-              "--min-resync-period=3m",
-              "--v=2"
+        "/hyperkube",
+        "controller-manager",
+        "--master=127.0.0.1:8080",
```
Comment (cheld) [inline cluster/images/hyperkube/static-pods/master-multi.json]: use kubeconfig instead for consistency reasons?

File: cluster/images/hyperkube/static-pods/master-multi.json [hunk 4]
```diff
@@ -57,18 +59,19 @@
...
       "image": "REGISTRY/hyperkube-ARCH:VERSION",
       "command": [
-              "/hyperkube",
-              "scheduler",
-              "--master=127.0.0.1:8080",
-              "--v=2"
-        ]
+        "/hyperkube",
+        "scheduler",
+        "--master=127.0.0.1:8080",
```
Comment (cheld) [inline cluster/images/hyperkube/static-pods/master-multi.json]: again, maybe kubeconfig?

File: cluster/images/hyperkube/static-pods/master-multi.json (hunk 3 above)
Comment (luxas) [inline cluster/images/hyperkube/static-pods/master-multi.json]: kubeconfig are mostly for the workers, but it might make sense to use it everywhere.
Also, I think we could expose the certs to the host via hostPath, don't you think?

File: cluster/images/hyperkube/static-pods/master-multi.json [hunk 5]
```diff
@@ -12,13 +12,15 @@
...
-              "--root-ca-file=/srv/kubernetes/ca.crt",
-              "--min-resync-period=3m",
-              "--v=2"
+        "/hyperkube",
+        "controller-manager",
+        "--master=127.0.0.1:8080",
+        "--service-account-private-key-file=/srv/kubernetes/server.key",
+        "--root-ca-file=/srv/kubernetes/ca.crt",
+        "--min-resync-period=3m",
+        "--leader-elect=true",
```
Comment (cheld) [inline cluster/images/hyperkube/static-pods/master-multi.json]: Out of curiosity, why did you add leader-elect?

File: cluster/images/hyperkube/static-pods/master-multi.json (hunk 3 above)
Comment (cheld) [inline cluster/images/hyperkube/static-pods/master-multi.json]: If I understand you correctly, you suggest to make /srv/kubernetes available on the host. So, e.g. a user could change the default password of admin. Sound good to me.

**Comment by cheld** [comment]
PTAL

**Comment by luxas** [comment]
PTAL @zreigz @cheld @mikedanese 
This one is for v1.4

**Comment by zreigz** [comment]
Today I will test it. Build and run e2e tests with the image

**Comment by luxas** [comment]
Thanks!

File: cluster/images/hyperkube/Dockerfile [hunk 6]
```diff
@@ -40,28 +38,22 @@ RUN cp /usr/bin/nsenter /nsenter
...
-COPY static-pods/etcd.json /etc/kubernetes/manifests/
-COPY static-pods/addon-manager.json /etc/kubernetes/manifests/
-
-# TODO: Move out kube-proxy to a DaemonSet again
-COPY static-pods/kube-proxy.json /etc/kubernetes/manifests/
+COPY static-pods/master.json \
+     static-pods/etcd.json \
+     static-pods/addon-manager-singlenode.json \
+     static-pods/kube-proxy.json \
+        /etc/kubernetes/manifests/
```
Comment (zreigz) [inline cluster/images/hyperkube/Dockerfile]: This is some limitation of Dockerfile engine because copying multiple files from subdirectory does not work.
```
Step 5 : COPY static-pods/master.json static-pods/etcd.json static-pods/addon-manager-singlenode.json static-pods/kube-proxy.json /etc/kubernetes/manifests/
lstat static-pods/master.json: no such file or directory
Makefile:56: recipe for target 'build' failed
```
Does it work on your side ?

File: cluster/images/hyperkube/Dockerfile [hunk 7]
```diff
@@ -40,28 +38,22 @@ RUN cp /usr/bin/nsenter /nsenter
...
 
 # Manifests for the docker-multinode guide
-COPY static-pods/master-multi.json /etc/kubernetes/manifests-multi/
-COPY static-pods/addon-manager.json /etc/kubernetes/manifests-multi/
-
-# TODO: Move out kube-proxy to a DaemonSet again
-COPY static-pods/kube-proxy.json /etc/kubernetes/manifests-multi/
+COPY static-pods/master-multi.json \
+     static-pods/addon-manager-multinode.json \
+        /etc/kubernetes/manifests-multi/
```
Comment (zreigz) [inline cluster/images/hyperkube/Dockerfile]: The same problem here.

File: cluster/images/hyperkube/Dockerfile (hunk 6 above)
Comment (zreigz) [inline cluster/images/hyperkube/Dockerfile]: Those lines are not a problem :) The real problem is here: https://github.com/kubernetes/kubernetes/pull/30299/files#diff-86f22fb5b693949f04077054c41b762fR80
You removed all json files :)

File: cluster/images/hyperkube/Makefile [hunk 8]
```diff
@@ -54,60 +54,58 @@ ifndef VERSION
...
 	cd ${TEMP_DIR} && sed -i.back "s|ARCH|${QEMUARCH}|g" Dockerfile
 	cd ${TEMP_DIR} && sed -i.back "s|BASEIMAGE|${BASEIMAGE}|g" Dockerfile
-	cd ${TEMP_DIR} && sed -i.back "s|-amd64|-${ARCH}|g" addons/*.yaml
-	cd ${TEMP_DIR} && sed -i.back "s|__PILLAR__DNS__REPLICAS__|1|g;s|__PILLAR__DNS__SERVER__|10.0.0.10|g;" addons/skydns*.yaml
-	cd ${TEMP_DIR} && sed -i.back "s|__PILLAR__DNS__DOMAIN__|cluster.local|g;s|__PILLAR__FEDERATIONS__DOMAIN__MAP__||g;" addons/skydns*.yaml
-	rm ${TEMP_DIR}/addons/*.back
+	cd ${TEMP_DIR} && sed -i.back "s|-amd64|-${ARCH}|g" addons/singlenode/*.yaml addons/multinode/*.yaml
+	cd ${TEMP_DIR} && sed -i.back "s|__PILLAR__DNS__REPLICAS__|1|g;s|__PILLAR__DNS__SERVER__|10.0.0.10|g;" addons/singlenode/skydns*.yaml addons/multinode/skydns*.yaml
+	cd ${TEMP_DIR} && sed -i.back "s|__PILLAR__DNS__DOMAIN__|cluster.local|g;s|__PILLAR__FEDERATIONS__DOMAIN__MAP__||g;" addons/singlenode/skydns*.yaml addons/multinode/skydns*.yaml
+	cd ${TEMP_DIR} && rm -f addons/singlenode/*.back addons/multinode/*.back static-pods/*.json
```
Comment (zreigz) [inline cluster/images/hyperkube/Makefile]: should be `cd ${TEMP_DIR} && rm -f addons/singlenode/*.back addons/multinode/*.back`

**Comment by zreigz** [comment]
I've created hyperkube image and I have error for apiserver
```
W0823 09:07:04.033692       1 server.go:173] No RSA key provided, service account token authentication disabled
F0823 09:07:04.033712       1 server.go:206] Invalid Authentication Config: open /srv/kubernetes/basic_auth.csv: no such file or directory
```

**Comment by zreigz** [comment]
PTAL

File: cluster/images/hyperkube/Makefile (hunk 8 above)
Comment (cheld) [inline cluster/images/hyperkube/Makefile]: I think it should be

rm -f addons/singlenode/*.back addons/multinode/*.back static-pods/*.back

File: cluster/images/hyperkube/static-pods/master-multi.json (hunk 2 above)
Comment (luxas) [inline cluster/images/hyperkube/static-pods/master-multi.json]: @zreigz The problem is here with your apiserver not starting. Somehow `/data` is still left here, but t should certainly be `/srv/kubernetes`

**Comment by luxas** [comment]
Updated the two unfortunate typos :)

**Comment by zreigz** [comment]
Docker build - works
hypercube - works
I couldn't start e2e tests. It hangs. I also notice you used old version of k8s in your branch: `version.Info{Major:"1", Minor:"2+", GitVersion:"v1.2.0-alpha.4.12658+1408772d738e7b",` 
is it intentionally ? If not maybe it would be good to rebase this branch.
Code looks good for me but I am not sure about functionality without e2e tests.

**Comment by luxas** [comment]
This fully works now :)
Both docker and docker-multinode.
Also rebased.

ping @mikedanese 
Now kube-proxy is in a daemonset

**Comment by luxas** [comment]
@mikedanese @david-mcmahon I'd really appreciate if we could get this into `v1.4.0-alpha.3`.
Can you take a quick look and LGTM?

**Comment by david-mcmahon** [comment]
LGTM.

**Comment by luxas** [comment]
@david-mcmahon Had to update one line that was slightly wrong, but now it should be okay.
Reapply the label when the bot is complaining, no huge changes were made

**Comment by pwittrock** [comment]
@luxas is 1.4 the correct milestone for this?  Is this a new feature or a bug fix?  Does something in 1.4 require this?
